contracts/infrastructure/storage/TokenPriceStorage.sol

```diff
@@ -23,32 +23,61 @@ import "../base/Managed.sol";
  * The contract only defines basic setters and getters with no logic.
  * Only managers of this contract can modify its state.
  */
 contract TokenPriceStorage is ITokenPriceStorage, Storage, Managed {
-    mapping(address => uint256) public cachedPrices;
-
-    function getTokenPrice(address _token) external override view returns (uint256 _price) {
-        _price = cachedPrices[_token];
+    struct TokenInfo {
+        uint184 cachedPrice;
+        uint64 updatedAt;
+        bool isTradable;
     }
 
-    function getPriceForTokenList(address[] calldata _tokens) external override view returns (uint256[] memory) {
-        uint256[] memory prices = new uint256[](_tokens.length);
+    // Price info per token
+    mapping(address => TokenInfo) public tokenInfo;
+    // The minimum period between two price updates
+    uint256 public minPriceUpdatePeriod;
+
+
+    // Getters
+
+    function getTokenPrice(address _token) external override view returns (uint184 _price) {
+        _price = tokenInfo[_token].cachedPrice;
+    }
+    function isTokenTradable(address _token) external override view returns (bool _isTradable) {
+        _isTradable = tokenInfo[_token].isTradable;
+    }
+    function getPriceForTokenList(address[] calldata _tokens) external view returns (uint184[] memory _prices) {
+        _prices = new uint184[](_tokens.length);
         for (uint i = 0; i < _tokens.length; i++) {
-            prices[i] = cachedPrices[_tokens[i]];
+            _prices[i] = tokenInfo[_tokens[i]].cachedPrice;
+        }
+    }
+    function getTradableForTokenList(address[] calldata _tokens) external view returns (bool[] memory _tradable) {
+        _tradable = new bool[](_tokens.length);
+        for (uint i = 0; i < _tokens.length; i++) {
+            _tradable[i] = tokenInfo[_tokens[i]].isTradable;
         }
-        return prices;
     }
 
-    function setPriceForTokenList(address[] calldata _tokens, uint256[] calldata _prices) external override onlyManager {
+    // Setters
+    
+    function setMinPriceUpdatePeriod(uint256 _newPeriod) external onlyOwner {
+        minPriceUpdatePeriod = _newPeriod;
+    }
+    function setPriceForTokenList(address[] calldata _tokens, uint184[] calldata _prices) external onlyManager {
         for (uint16 i = 0; i < _tokens.length; i++) {
-            cachedPrices[_tokens[i]] = _prices[i];
+            uint64 updatedAt = tokenInfo[_tokens[i]].updatedAt;
+            require(updatedAt == 0 || block.timestamp >= updatedAt + minPriceUpdatePeriod, "TPS: Price updated too early");
+            tokenInfo[_tokens[i]].cachedPrice = _prices[i];
+            tokenInfo[_tokens[i]].updatedAt = uint64(block.timestamp);
         }
     }
-
-    function setPrice(address _token, uint256 _price) external override onlyManager {
-        cachedPrices[_token] = _price;
+    function setTradableForTokenList(address[] calldata _tokens, bool[] calldata _tradable) external {
+        for (uint16 i = 0; i < _tokens.length; i++) {
```

Comment:
Same comment as above: we must check that the arrays have equal length and probably use a `uint256` for the loop.

Also, why are we preventing `managers` to make tokens tradable? I think it is ok to let managers make tokens tradable, and it will make it much easier operationally.

Reply:
If a manager could set a token to `tradable` then there would be a possibility for a compromised manager to immediately add a fake token and set `tradable` to `true` to bypass the daily limit and drain any wallet it controls. 

Since we decided to introduce `minPriceUpdatePeriod` to mitigate the power of a manager, I think it makes sense to also restrict a manager's ability to add tradable tokens.

Reply:
I personally think that the added (operational) complexity is not worth the risk. Having to go through the multisig every-time we want to add a new token will be a pain, while the risk of someone compromising a wallet AND compromising our backend is minimal.

Reply:
Ok ignore my last comment, you are right.